korgan/kaspi_ofd.py

```
@@ -0,0 +1,393 @@
+        if "hash" in params and params["hash"] and not re.fullmatch(r"[0-9a-fA-F]{32,128}", params["hash"]):
+            raise KaspiOFDVerificationError("Некорректная подпись ссылки Kaspi ОФД")
+        ordered = [(key, params[key]) for key in ("extTranId", "hash", "locale", "sale_date") if params.get(key)]
+        if not any(key == "locale" for key, _ in ordered):
+            ordered.insert(-1 if ordered else 0, ("locale", "ru"))
+        return urlunparse(("https", _KASPI_RECEIPT_HOST, _KASPI_DOWNLOAD_PATH, "", urlencode(ordered), ""))
+
+    if path == _KASPI_FISCAL_PATH:
+        params = _query_dict(parsed, {"f", "i", "s", "t", "locale"})
+        for required in ("f", "i", "s", "t"):
+            if not params.get(required):
+                raise KaspiOFDVerificationError("В QR отсутствуют обязательные фискальные данные")
```

Ruff (0.16.2)

[warning] Avoid specifying long messages outside the exception class

(TRY003)